File: scimesh/sdk/schema.py
"""Bounded JSON Schema subset used for SDK v1 public parameters."""

from __future__ import annotations

import math
import re
from fractions import Fraction
from typing import Mapping


_ANNOTATIONS = {
    "$schema",
    "title",
    "description",
    "default",
    "examples",
    "deprecated",
    "readOnly",
    "writeOnly",
}
_KEYWORDS = _ANNOTATIONS | {
    "type",
    "enum",
    "const",
    "properties",
    "additionalProperties",
    "required",
    "minProperties",
    "maxProperties",
    "items",
    "minItems",
    "maxItems",
    "uniqueItems",
    "minLength",
    "maxLength",
    "pattern",
    "minimum",
    "maximum",
    "exclusiveMinimum",
    "exclusiveMaximum",
    "multipleOf",
    "allOf",
    "anyOf",
    "oneOf",
    "not",
}
_TYPES = {"null", "boolean", "object", "array", "number", "integer", "string"}


class ParameterValidationError(ValueError):
    """Sanitized public-parameter schema failure."""


def _schema_error(message: str) -> ValueError:
    return ValueError("unsupported or invalid parameters_schema: " + message)


def _json_equal(left: object, right: object) -> bool:
    """Compare values using the JSON data model rather than Python coercion.

    Python considers ``True == 1`` while JSON has distinct boolean and number
    types. JSON Schema does, however, treat integral and non-integral syntax for
    the same mathematical number (for example ``1`` and ``1.0``) as equal.
    """
    if isinstance(left, bool) or isinstance(right, bool):
        return isinstance(left, bool) and isinstance(right, bool) and left is right
    if isinstance(left, (int, float)) and isinstance(right, (int, float)):
        return left == right
    if left is None or right is None:
        return left is None and right is None
    if isinstance(left, str) or isinstance(right, str):
        return isinstance(left, str) and isinstance(right, str) and left == right
    if isinstance(left, Mapping) and isinstance(right, Mapping):
        return set(left) == set(right) and all(
            _json_equal(left[key], right[key]) for key in left
        )
    if isinstance(left, (list, tuple)) and isinstance(right, (list, tuple)):
        return len(left) == len(right) and all(
            _json_equal(left_item, right_item)
            for left_item, right_item in zip(left, right)
        )
    return False


def _json_key(value: object, depth: int = 0) -> object:
    """Build a hashable JSON-type-aware key in linear time."""
    if depth > 64:
        raise ValueError("JSON value nesting exceeds 64 levels")
    if value is None:
        return ("null",)
    if isinstance(value, bool):
        return ("boolean", value)
    if isinstance(value, (int, float)):
        return (
            "number",
            Fraction(value) if isinstance(value, int) else Fraction.from_float(value),
        )
    if isinstance(value, str):
        return ("string", value)
    if isinstance(value, Mapping):
        return (
            "object",
            tuple(
                (key, _json_key(child, depth + 1))
                for key, child in sorted(value.items())
            ),
        )
    if isinstance(value, (list, tuple)):
        return ("array", tuple(_json_key(child, depth + 1) for child in value))
    raise ValueError("value is not JSON-compatible")


def _validate_safe_pattern(pattern: str) -> None:
    """Accept only the v1 linear-time regex subset.

    Groups, alternation, backreferences, and repetition operators are excluded;
    literals, anchors, character classes, escapes, and ``.`` remain available.
    """
    escaped = False
    in_class = False
    for character in pattern:
        if escaped:
            if character.isdigit():
                raise _schema_error("pattern backreferences are not supported")
            escaped = False
            continue
        if character == "\\":
            escaped = True
            continue
        if character == "[" and not in_class:
            in_class = True
            continue
        if character == "]" and in_class:
            in_class = False
            continue
        if not in_class and character in "()|*+?{}":
            raise _schema_error("pattern uses an unbounded regex operator")
    if escaped or in_class:
        # ``re.compile`` will provide the canonical invalid-regex error below.
        return


def _is_json_multiple(value: int | float, divisor: int | float) -> bool:
    """Evaluate ``multipleOf`` without converting arbitrary integers to float."""
    if isinstance(value, int) and isinstance(divisor, int):
        return value % divisor == 0
    value_fraction = Fraction(value) if isinstance(value, int) else Fraction(str(value))
    divisor_fraction = (
        Fraction(divisor) if isinstance(divisor, int) else Fraction(str(divisor))
    )
    return (value_fraction / divisor_fraction).denominator == 1


def validate_schema_definition(
    schema: Mapping[str, object], *, _depth: int = 0
) -> None:
    """Validate a parameter schema against the bounded JSON Schema subset.

    Raises ``ValueError`` on unknown keywords, unsupported types, unsafe
    patterns, or malformed bounds.
    """
    if _depth > 64:
        raise _schema_error("nesting exceeds 64 levels")
    if not isinstance(schema, Mapping):
        raise _schema_error("each schema node must be an object")
    unknown = set(schema) - _KEYWORDS
    if unknown:
        raise _schema_error("unknown keyword " + sorted(unknown)[0])
    raw_type = schema.get("type")
    if raw_type is not None:
        declared = (raw_type,) if isinstance(raw_type, str) else raw_type
        if not isinstance(declared, (list, tuple)) or not declared:
            raise _schema_error("type must be a string or non-empty array")
        if any(not isinstance(value, str) or value not in _TYPES for value in declared):
            raise _schema_error("type contains an unsupported JSON type")
        if len(declared) != len(set(declared)):
            raise _schema_error("type alternatives must be unique")
    properties = schema.get("properties")
    if properties is not None:
        if not isinstance(properties, Mapping) or any(
            not isinstance(name, str) for name in properties
        ):
            raise _schema_error("properties must be an object")
        for child in properties.values():
            validate_schema_definition(child, _depth=_depth + 1)  # type: ignore[arg-type]
    additional = schema.get("additionalProperties")
    if additional is not None and not isinstance(additional, (bool, Mapping)):
        raise _schema_error("additionalProperties must be a boolean or schema")
    if isinstance(additional, Mapping):
        validate_schema_definition(additional, _depth=_depth + 1)
    required = schema.get("required")
    if required is not None:
        if not isinstance(required, (list, tuple)) or any(
            not isinstance(name, str) for name in required
        ):
            raise _schema_error("required must be an array of strings")
        if len(required) != len(set(required)):
            raise _schema_error("required names must be unique")
    for keyword in ("items", "not"):
        child = schema.get(keyword)
        if child is not None:
            validate_schema_definition(child, _depth=_depth + 1)  # type: ignore[arg-type]
    for keyword in ("allOf", "anyOf", "oneOf"):
        children = schema.get(keyword)
        if children is None:
            continue
        if not isinstance(children, (list, tuple)) or not children:
            raise _schema_error(f"{keyword} must be a non-empty array")
        for child in children:
            validate_schema_definition(child, _depth=_depth + 1)  # type: ignore[arg-type]
    enum = schema.get("enum")
    if enum is not None and (not isinstance(enum, (list, tuple)) or not enum):
        raise _schema_error("enum must be a non-empty array")
    if isinstance(enum, (list, tuple)):
        seen_enum: set[object] = set()
        for item in enum:
            key = _json_key(item)
            if key in seen_enum:
                raise _schema_error("enum values must be unique")
            seen_enum.add(key)
    for keyword in (
        "minProperties",
        "maxProperties",
        "minItems",
        "maxItems",
        "minLength",
        "maxLength",
    ):
        value = schema.get(keyword)
        if value is not None and (
            isinstance(value, bool) or not isinstance(value, int) or value < 0
        ):
            raise _schema_error(f"{keyword} must be a non-negative integer")
    for minimum, maximum in (
        ("minProperties", "maxProperties"),
        ("minItems", "maxItems"),
        ("minLength", "maxLength"),
    ):
        if (
            minimum in schema
            and maximum in schema
            and schema[minimum] > schema[maximum]  # type: ignore[operator]
        ):
            raise _schema_error(f"{minimum} must not exceed {maximum}")
    for keyword in (
        "minimum",
        "maximum",
        "exclusiveMinimum",
        "exclusiveMaximum",
        "multipleOf",
    ):
        value = schema.get(keyword)
        if value is not None and (
            isinstance(value, bool)
            or not isinstance(value, (int, float))
            or (isinstance(value, float) and not math.isfinite(value))
        ):
            raise _schema_error(f"{keyword} must be a finite number")
    if "multipleOf" in schema and schema["multipleOf"] <= 0:  # type: ignore[operator]
        raise _schema_error("multipleOf must be positive")
    pattern = schema.get("pattern")
    if pattern is not None:
        if not isinstance(pattern, str) or len(pattern) > 1024:
            raise _schema_error("pattern must be a string of at most 1024 characters")
        _validate_safe_pattern(pattern)
        try:
            re.compile(pattern)
        except re.error as error:
            raise _schema_error("pattern is not a valid regular expression") from error
    for keyword in ("uniqueItems", "deprecated", "readOnly", "writeOnly"):
        if keyword in schema and not isinstance(schema[keyword], bool):
            raise _schema_error(f"{keyword} must be a boolean")


def _type_matches(value: object, expected: str) -> bool:
    if expected == "null":
        return value is None
    if expected == "boolean":
        return isinstance(value, bool)
    if expected == "object":
        return isinstance(value, Mapping)
    if expected == "array":
        return isinstance(value, (list, tuple))
    if expected == "integer":
        return isinstance(value, int) and not isinstance(value, bool)
    if expected == "number":
        return isinstance(value, (int, float)) and not isinstance(value, bool)
    if expected == "string":
        return isinstance(value, str)
    return False


def _failure(path: str, reason: str) -> ParameterValidationError:
    return ParameterValidationError(
        f"job parameters violate their schema at {path}: {reason}"
    )


def validate_parameter_instance(
    value: object,
    schema: Mapping[str, object],
    *,
    path: str = "$",
    _depth: int = 0,
) -> None:
    """Validate one parameter value against a schema subset node.

    Raises ``ParameterValidationError`` (a ``ValueError``) with a sanitized
    JSON path on the first violation.
    """
    if _depth > 64:
        raise _failure(path, "nesting exceeds 64 levels")
    raw_type = schema.get("type")
    if raw_type is not None:
        expected = (raw_type,) if isinstance(raw_type, str) else tuple(raw_type)  # type: ignore[arg-type]
        if not any(_type_matches(value, item) for item in expected):
            raise _failure(path, "type mismatch")
    if "enum" in schema and not any(
        _json_equal(value, candidate)
        for candidate in schema["enum"]  # type: ignore[union-attr]
    ):
        raise _failure(path, "value is outside enum")
    if "const" in schema and not _json_equal(value, schema["const"]):
        raise _failure(path, "value does not match const")
    for keyword in ("allOf", "anyOf", "oneOf"):
        children = schema.get(keyword)
        if children is None:
            continue
        matches = 0
        for child in children:  # type: ignore[union-attr]
            try:
                validate_parameter_instance(value, child, path=path, _depth=_depth + 1)
            except ParameterValidationError:
                continue
            matches += 1
        if keyword == "allOf" and matches != len(children):  # type: ignore[arg-type]
            raise _failure(path, "allOf did not match")
        if keyword == "anyOf" and matches == 0:
            raise _failure(path, "anyOf did not match")
        if keyword == "oneOf" and matches != 1:
            raise _failure(path, "oneOf did not match exactly once")
    excluded = schema.get("not")
    if excluded is not None:
        try:
            validate_parameter_instance(value, excluded, path=path, _depth=_depth + 1)  # type: ignore[arg-type]
        except ParameterValidationError:
            pass
        else:
            raise _failure(path, "value matches a forbidden schema")
    if isinstance(value, Mapping):
        required = schema.get("required", ())
        missing = set(required) - set(value)  # type: ignore[arg-type]
        if missing:
            raise _failure(path, "missing required field " + sorted(missing)[0])
        minimum = schema.get("minProperties")
        maximum = schema.get("maxProperties")
        if minimum is not None and len(value) < minimum:  # type: ignore[operator]
            raise _failure(path, "too few properties")
        if maximum is not None and len(value) > maximum:  # type: ignore[operator]
            raise _failure(path, "too many properties")
        properties = schema.get("properties", {})
        additional = schema.get("additionalProperties", True)
        for name, child in value.items():
            if name in properties:  # type: ignore[operator]
                validate_parameter_instance(
                    child,
                    properties[name],  # type: ignore[index]
                    path=f"{path}.{name}",
                    _depth=_depth + 1,
                )
            elif additional is False:
                raise _failure(path, f"unknown field {name}")
            elif isinstance(additional, Mapping):
                validate_parameter_instance(
                    child, additional, path=f"{path}.{name}", _depth=_depth + 1
                )
    if isinstance(value, (list, tuple)):
        minimum = schema.get("minItems")
        maximum = schema.get("maxItems")
        if minimum is not None and len(value) < minimum:  # type: ignore[operator]
            raise _failure(path, "too few items")
        if maximum is not None and len(value) > maximum:  # type: ignore[operator]
            raise _failure(path, "too many items")
        if schema.get("uniqueItems"):
            seen_items: set[object] = set()
            for item in value:
                key = _json_key(item)
                if key in seen_items:
                    raise _failure(path, "items must be unique")
                seen_items.add(key)
        child_schema = schema.get("items")
        if child_schema is not None:
            for index, item in enumerate(value):
                validate_parameter_instance(
                    item,
                    child_schema,  # type: ignore[arg-type]
                    path=f"{path}[{index}]",
                    _depth=_depth + 1,
                )
    if isinstance(value, str):
        if "minLength" in schema and len(value) < schema["minLength"]:  # type: ignore[operator]
            raise _failure(path, "string is too short")
        if "maxLength" in schema and len(value) > schema["maxLength"]:  # type: ignore[operator]
            raise _failure(path, "string is too long")
        if "pattern" in schema and re.search(schema["pattern"], value) is None:  # type: ignore[arg-type]
            raise _failure(path, "string does not match pattern")
    if isinstance(value, (int, float)) and not isinstance(value, bool):
        checks = (
            ("minimum", lambda actual, bound: actual >= bound),
            ("maximum", lambda actual, bound: actual <= bound),
            ("exclusiveMinimum", lambda actual, bound: actual > bound),
            ("exclusiveMaximum", lambda actual, bound: actual < bound),
        )
        for keyword, predicate in checks:
            if keyword in schema and not predicate(value, schema[keyword]):
                raise _failure(path, f"number violates {keyword}")
        if "multipleOf" in schema:
            if not _is_json_multiple(value, schema["multipleOf"]):  # type: ignore[arg-type]
                raise _failure(path, "number violates multipleOf")

File: coordinator/internal/workloads/schema.go
package workloads

import (
	"fmt"
	"math"
	"sort"
)

// validateSchema checks the strict JSON Schema subset used by SDK manifests.
// It mirrors what the SDK registry enforces on the Python side: an object
// schema with additionalProperties=false, typed properties, and the keyword
// subset the coordinator understands (type, enum, required, minimum/maximum,
// minLength/maxLength, oneOf, not).
func validateSchema(workloadName string, schema map[string]any) error {
	if schema == nil {
		return fmt.Errorf("workload %q has no parameter schema", workloadName)
	}
	if err := validateSchemaNode(workloadName+".parameters_schema", schema); err != nil {
		return err
	}
	if schemaType(schema) != "object" {
		return fmt.Errorf("workload %q parameter schema must be an object schema", workloadName)
	}
	if additional, ok := schema["additionalProperties"].(bool); !ok || additional {
		return fmt.Errorf("workload %q parameter schema must set additionalProperties=false", workloadName)
	}
	properties, ok := schema["properties"].(map[string]any)
	if !ok {
		return fmt.Errorf("workload %q parameter schema must declare properties", workloadName)
	}
	for name, property := range properties {
		child, ok := property.(map[string]any)
		if !ok {
			return fmt.Errorf("workload %q parameter %q must be a schema object", workloadName, name)
		}
		if err := validateSchemaNode(name, child); err != nil {
			return err
		}
	}
	return nil
}

func validateSchemaNode(field string, node map[string]any) error {
	for keyword := range node {
		switch keyword {
		case "type", "enum", "required", "minimum", "maximum", "exclusiveMinimum",
			"exclusiveMaximum", "minLength", "maxLength", "properties",
			"additionalProperties", "oneOf", "not", "default", "description",
			"items", "minItems", "maxItems":
		default:
			return fmt.Errorf("%s uses unsupported JSON Schema keyword %q", field, keyword)
		}
	}
	if rawType, ok := node["type"]; ok {
		schemaType, ok := rawType.(string)
		if !ok {
			return fmt.Errorf("%s type must be a string", field)
		}
		switch schemaType {
		case "string", "number", "integer", "boolean", "object", "array":
		default:
			return fmt.Errorf("%s has unknown type %q", field, schemaType)
		}
	}
	for _, keyword := range []string{"minimum", "maximum", "exclusiveMinimum", "exclusiveMaximum"} {
		if value, ok := node[keyword]; ok {
			if number, ok := value.(float64); !ok || math.IsNaN(number) || math.IsInf(number, 0) {
				return fmt.Errorf("%s %s must be a finite number", field, keyword)
			}
		}
	}
	for _, keyword := range []string{"minLength", "maxLength", "minItems", "maxItems"} {
		if value, ok := node[keyword]; ok {
			if number, ok := value.(float64); !ok || number < 0 || number != math.Trunc(number) {
				return fmt.Errorf("%s %s must be a non-negative integer", field, keyword)
			}
		}
	}
	if required, ok := node["required"]; ok {
		entries, ok := required.([]any)
		if !ok {
			return fmt.Errorf("%s required must be an array of strings", field)
		}
		for _, entry := range entries {
			if _, ok := entry.(string); !ok {
				return fmt.Errorf("%s required must be an array of strings", field)
			}
		}
	}
	if enums, ok := node["enum"]; ok {
		entries, ok := enums.([]any)
		if !ok || len(entries) == 0 {
			return fmt.Errorf("%s enum must be a non-empty array", field)
		}
	}
	if oneOf, ok := node["oneOf"]; ok {
		entries, ok := oneOf.([]any)
		if !ok || len(entries) == 0 {
			return fmt.Errorf("%s oneOf must be a non-empty array", field)
		}
		for index, entry := range entries {
			child, ok := entry.(map[string]any)
			if !ok {
				return fmt.Errorf("%s oneOf[%d] must be a schema object", field, index)
			}
			if err := validateSchemaNode(fmt.Sprintf("%s.oneOf[%d]", field, index), child); err != nil {
				return err
			}
		}
	}
	if child, ok := node["not"]; ok {
		not, ok := child.(map[string]any)
		if !ok {
			return fmt.Errorf("%s not must be a schema object", field)
		}
		if err := validateSchemaNode(field+".not", not); err != nil {
			return err
		}
	}
	if items, ok := node["items"]; ok {
		child, ok := items.(map[string]any)
		if !ok {
			return fmt.Errorf("%s items must be a schema object", field)
		}
		if err := validateSchemaNode(field+".items", child); err != nil {
			return err
		}
	}
	if properties, ok := node["properties"]; ok {
		entries, ok := properties.(map[string]any)
		if !ok {
			return fmt.Errorf("%s properties must be an object", field)
		}
		for name, property := range entries {
			child, ok := property.(map[string]any)
			if !ok {
				return fmt.Errorf("%s property %q must be a schema object", field, name)
			}
			if err := validateSchemaNode(field+"."+name, child); err != nil {
				return err
			}
		}
	}
	return nil
}

func schemaType(node map[string]any) string {
	rawType, _ := node["type"].(string)
	return rawType
}

// validateParameters checks values against the strict schema subset.
func validateParameters(workloadName string, schema map[string]any, parameters map[string]any) error {
	properties, _ := schema["properties"].(map[string]any)
	for name := range parameters {
		if _, declared := properties[name]; !declared {
			return fmt.Errorf("workload %q does not accept parameter %q", workloadName, name)
		}
	}
	required, _ := schema["required"].([]any)
	for _, name := range required {
		field, _ := name.(string)
		if _, present := parameters[field]; !present {
			return fmt.Errorf("workload %q requires parameter %q", workloadName, field)
		}
	}
	for name, value := range parameters {
		property, declared := properties[name].(map[string]any)
		if !declared {
			continue
		}
		if err := validateProperty(workloadName+"."+name, property, value); err != nil {
			return err
		}
	}
	if oneOf, ok := schema["oneOf"].([]any); ok && len(oneOf) > 0 {
		if err := validateOneOf(workloadName, oneOf, parameters); err != nil {
			return err
		}
	}
	return nil
}

func validateProperty(field string, property map[string]any, value any) error {
	if enums, ok := property["enum"].([]any); ok {
		for _, candidate := range enums {
			if valuesEqual(candidate, value) {
				return nil
			}
		}
		return fmt.Errorf("%s must be one of the declared enum values", field)
	}
	switch schemaType(property) {
	case "string":
		text, ok := value.(string)
		if !ok {
			return fmt.Errorf("%s must be a string", field)
		}
		if minimum, ok := lengthBound(property["minLength"]); ok && len([]rune(text)) < minimum {
			return fmt.Errorf("%s is shorter than the minimum length", field)
		}
		if maximum, ok := lengthBound(property["maxLength"]); ok && len([]rune(text)) > maximum {
			return fmt.Errorf("%s exceeds the maximum length", field)
		}
	case "number", "integer":
		number, ok := asFloat(value)
		if !ok {
			return fmt.Errorf("%s must be a number", field)
		}
		if schemaType(property) == "integer" && number != math.Trunc(number) {
			return fmt.Errorf("%s must be an integer", field)
		}
		if minimum, ok := numberBound(property["minimum"]); ok && number < minimum {
			return fmt.Errorf("%s is below the minimum", field)
		}
		if maximum, ok := numberBound(property["maximum"]); ok && number > maximum {
			return fmt.Errorf("%s exceeds the maximum", field)
		}
	case "boolean":
		if _, ok := value.(bool); !ok {
			return fmt.Errorf("%s must be a boolean", field)
		}
	case "object":
		child, ok := value.(map[string]any)
		if !ok {
			return fmt.Errorf("%s must be an object", field)
		}
		properties, _ := property["properties"].(map[string]any)
		for name := range child {
			if _, declared := properties[name]; !declared {
				return fmt.Errorf("%s has undeclared field %q", field, name)
			}
		}
	case "array":
		items, ok := value.([]any)
		if !ok {
			return fmt.Errorf("%s must be an array", field)
		}
		if itemSchema, ok := property["items"].(map[string]any); ok {
			for index, item := range items {
				if err := validateProperty(fmt.Sprintf("%s[%d]", field, index), itemSchema, item); err != nil {
					return err
				}
			}
		}
	case "":
		// No type keyword: enum-only properties are handled above.
		return fmt.Errorf("%s has no JSON Schema type", field)
	}
	return nil
}

func validateOneOf(workloadName string, oneOf []any, parameters map[string]any) error {
	satisfied := 0
	for _, candidate := range oneOf {
		option, ok := candidate.(map[string]any)
		if !ok {
			continue
		}
		if optionSatisfied(option, parameters) {
			satisfied++
		}
	}
	if satisfied != 1 {
		return fmt.Errorf("workload %q requires exactly one of the declared parameter alternatives", workloadName)
	}
	return nil
}

func optionSatisfied(option map[string]any, parameters map[string]any) bool {
	if required, ok := option["required"].([]any); ok {
		for _, name := range required {
			field, _ := name.(string)
			if _, present := parameters[field]; !present {
				return false
			}
		}
	}
	if not, ok := option["not"].(map[string]any); ok {
		if required, ok := not["required"].([]any); ok {
			for _, name := range required {
				field, _ := name.(string)
				if _, present := parameters[field]; present {
					return false
				}
			}
		}
	}
	return true
}

func lengthBound(value any) (int, bool) {
	number, ok := value.(float64)
	if !ok || number != math.Trunc(number) {
		return 0, false
	}
	return int(number), true
}

func numberBound(value any) (float64, bool) {
	number, ok := asFloat(value)
	if !ok || math.IsNaN(number) || math.IsInf(number, 0) {
		return 0, false
	}
	return number, true
}

func asFloat(value any) (float64, bool) {
	switch v := value.(type) {
	case float64:
		return v, true
	case float32:
		return float64(v), true
	case int:
		return float64(v), true
	case int32:
		return float64(v), true
	case int64:
		return float64(v), true
	}
	return 0, false
}

func valuesEqual(left, right any) bool {
	switch l := left.(type) {
	case float64:
		r, ok := right.(float64)
		return ok && l == r
	case string:
		r, ok := right.(string)
		return ok && l == r
	case bool:
		r, ok := right.(bool)
		return ok && l == r
	case nil:
		return right == nil
	}
	return false
}

// schemaDefaults collects the declared default for each property. The UI uses
// these to pre-fill controls that have no workload-declared UI default.
func schemaDefaults(schema map[string]any) map[string]any {
	properties, _ := schema["properties"].(map[string]any)
	defaults := map[string]any{}
	for name, property := range properties {
		child, ok := property.(map[string]any)
		if !ok {
			continue
		}
		if value, present := child["default"]; present {
			defaults[name] = value
		}
	}
	return defaults
}

// SortedFields returns the sorted declared parameter names.
func SortedFields(schema map[string]any) []string {
	properties, _ := schema["properties"].(map[string]any)
	names := make([]string, 0, len(properties))
	for name := range properties {
		names = append(names, name)
	}
	sort.Strings(names)
	return names
}
